File: src/com/android/settings/edith/dashboard/RestrictedHomepagePreference.java
package com.android.settings.edith.dashboard;

import android.content.Context;
import android.util.AttributeSet;
import android.view.View;

import androidx.preference.PreferenceViewHolder;

import com.android.settings.R;
import com.android.settingslib.RestrictedTopLevelPreference;

/**
 * Homepage preference that supports device admin / user restriction.
 */
public class RestrictedHomepagePreference extends RestrictedTopLevelPreference {

    private View mIcon;
    private View mText;
    private boolean mIconVisible = true;
    private int mIconPaddingStart = -1;
    private int mTextPaddingStart = -1;

    public RestrictedHomepagePreference(Context context, AttributeSet attrs, int defStyleAttr,
                                        int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        initLayout();
    }

    public RestrictedHomepagePreference(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initLayout();
    }

    public RestrictedHomepagePreference(Context context, AttributeSet attrs) {
        super(context, attrs);
        initLayout();
    }

    public RestrictedHomepagePreference(Context context) {
        super(context);
        initLayout();
    }

    private void initLayout() {
        setLayoutResource(R.layout.edith_homepage_preference);
    }

    public void setIconVisible(boolean visible) {
        mIconVisible = visible;
        if (mIcon != null) {
            mIcon.setVisibility(visible ? View.VISIBLE : View.GONE);
        }
    }

    public void setIconPaddingStart(int paddingStart) {
        mIconPaddingStart = paddingStart;
        if (mIcon != null && paddingStart >= 0) {
            mIcon.setPaddingRelative(paddingStart, mIcon.getPaddingTop(), mIcon.getPaddingEnd(),
                    mIcon.getPaddingBottom());
        }
    }

    public void setTextPaddingStart(int paddingStart) {
        mTextPaddingStart = paddingStart;
        if (mText != null && paddingStart >= 0) {
            mText.setPaddingRelative(paddingStart, mText.getPaddingTop(), mText.getPaddingEnd(),
                    mText.getPaddingBottom());
        }
    }

    @Override
    public void onBindViewHolder(PreferenceViewHolder holder) {
        super.onBindViewHolder(holder);
        mIcon = holder.findViewById(R.id.icon_frame);
        mText = holder.findViewById(R.id.text_frame);
        setIconVisible(mIconVisible);
        setIconPaddingStart(mIconPaddingStart);
        setTextPaddingStart(mTextPaddingStart);
    }
}

File: src/com/android/settings/edith/dashboard/RestrictedHomepagePreference.kt
/*
 * Copyright (C) 2026 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.settings.edith.dashboard

import android.content.Context
import android.util.AttributeSet
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.platform.ViewCompositionStrategy
import androidx.preference.PreferenceViewHolder
import com.android.settings.R
import com.android.settingslib.RestrictedTopLevelPreference
import com.android.settingslib.spa.framework.theme.SettingsTheme

/**
 * Homepage preference that supports device admin / user restriction.
 */
class RestrictedHomepagePreference @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0,
    defStyleRes: Int = 0,
) : RestrictedTopLevelPreference(context, attrs, defStyleAttr, defStyleRes) {

    private var _iconVisible by mutableStateOf(true)
    private var _iconPaddingStart by mutableIntStateOf(-1)
    private var _textPaddingStart by mutableIntStateOf(-1)

    init {
        layoutResource = R.layout.preference_compose
    }

    fun setIconVisible(visible: Boolean) {
        _iconVisible = visible
    }

    fun setIconPaddingStart(paddingStart: Int) {
        _iconPaddingStart = paddingStart
    }

    fun setTextPaddingStart(paddingStart: Int) {
        _textPaddingStart = paddingStart
    }

    override fun onBindViewHolder(holder: PreferenceViewHolder) {
        super.onBindViewHolder(holder)
        (holder.itemView as ComposeView).apply {
            setViewCompositionStrategy(
                ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed,
            )
            setContent {
                SettingsTheme {
                    EdithHomepageContent(
                        icon = icon,
                        title = title,
                        iconVisible = _iconVisible,
                        iconPaddingStartPx = _iconPaddingStart,
                        textPaddingStartPx = _textPaddingStart,
                    )
                }
            }
        }
    }
}
